subprojects/core/src/integTest/groovy/org/gradle/api/provider/ProviderIntegrationTest.groovy

```diff
@@ -0,0 +1,381 @@
+
+                public MyTask() {
+                    enabled = getProject().provider(false);
+                    outputFiles = getProject().provider(getProject().files());
+                }
+
+                @Input
+                public boolean getEnabled() {
+                    return enabled.get();
+                }
+                
+                public void setEnabled(Provider<Boolean> enabled) {
+                    this.enabled = enabled;
+                }
+                
+                @OutputFiles
+                public ConfigurableFileCollection getOutputFiles() {
+                    return outputFiles.get();
+                }
+
+                public void setOutputFiles(Provider<ConfigurableFileCollection> outputFiles) {
+                    this.outputFiles = outputFiles;
+                }
+                
+                @TaskAction
+                public void resolveValue() throws IOException {
+                    if (getEnabled()) {
+                        for (File outputFile : getOutputFiles()) {
+                            writeFile(outputFile, "$OUTPUT_FILE_CONTENT");
+                        }
+                    }
+                }
+                
+                private void writeFile(File destination, String content) throws IOException {
+                    BufferedWriter output = null;
+                    try {
+                        output = new BufferedWriter(new FileWriter(destination));
+                        output.write(content);
+                    } finally {
+                        if (output != null) {
+                            output.close();
+                        }
+                    }
+                }
+            }
+        """
+    }
+
+    private String pluginWithExtensionMapping(Closure taskCreation) {
+        """
+            apply plugin: MyPlugin
+            
+            pluginConfig {
+                enabled = true
+                outputFiles = files('${normaliseFileSeparators(customOutputFile.canonicalPath)}')
+            }
+
+            class MyPlugin implements Plugin<Project> {
+                void apply(Project project) {
+                    def extension = project.extensions.create('pluginConfig', MyExtension)
+
+                    ${taskCreation()}
+                }
+            }
+
+            class MyExtension {
+                Boolean enabled
+                FileCollection outputFiles
+            }
+        """
+    }
+
+    static String taskConfiguredWithConventionMapping() {
+        """
+            project.tasks.create('myTask', MyTask) {
+                conventionMapping.enabled = { extension.enabled }
+                conventionMapping.outputFiles = { extension.outputFiles }
+            }
+        """
+    }
+
+    static String taskConfiguredWithProvider() {
+        """
+            project.tasks.create('myTask', MyTask) {
+                enabled = project.provider { extension.enabled }
```

Comment:
I'd like to defer any convenience access for a later version if possible so we can release the functionality sooner.